### ISSUE TITLE
Tags ... not always working?

### ISSUE THREAD

**Issue by janboddez** [Automattic/wordpress-activitypub#667]
### Quick summary

I may be too quick (again ...) but tags don't seem to work anymore?

E.g., https://indieweb.social/@ochtendgrijs@mastodon.vlaanderen/111777445438425723: JSON shows `"tags":[]`.
The "same" post but crossposted (note that this copy did explicitly contain the hashtag in its content) over Mastodon's API shows: `"tags":[{"name":"auto","url":"https://indieweb.social/tags/auto"}]`.
The "original" over at https://ochtendgrijs.be/notes/65a0c00b88/?activitypub shows: `"tag":[{"type":"Hashtag","href":"https:\/\/ochtendgrijs.be\/tag\/auto\/","name":"#auto"}].`

### Steps to reproduce

1. Add a tag
2. Watch it being added to the local JSON
3. When viewed on Mastodon, the tag ... isn't there? Also, the post does not show up when searching for this tag.

### What you expected to happen

For posts to show up under the tag in question.

### What actually happened

_Looks like_ the tag isn't federating.

Update: But then for [this here post](https://indieweb.social/@ochtendgrijs@ochtendgrijs.be/111772184590065438), posted yesterday, it worked ...

Update no. 2: So I tried again, nope: https://indieweb.social/@jan@jan.boddez.net/111777535133712971.

### Impact

One

### Available workarounds?

No but the platform is still usable

**Comment by janboddez** [comment]
Dang, I think I found it.

The tag gets added as I update the post.

I'm suspecting the block editor is to blame here, ~~possibly~~ in combination with my caching strategy!

I use `save_post_post` and similar hooks to cache a copy of my post, but the block editor does not add the tags until later.

What I find so interesting is that the ActivityPub plugin does get away with `transition_post_status`, which runs at pretty much the same time.

It might be safer, going forward, if "block editor posts" used `rest_after_insert_post`. Either way, I think it still works because the use of `wp_schedule_single_event`: the event does not run until the next request (likely also block editor-initiated, where it is in use). In my particular case, it probably even gets kicked off by the "cache request"!

**Comment by janboddez** [comment]
Changed my caching script; doesn't work. I have to update the post explicitly to make a change in tags visible on Mastodon. Also merely adding an in-content hashtag does not necessarily tag the post. Need to add the corresponding tag (in WordPress) as well.

Still need to try without my "cache warming" script thingy.

**Comment by janboddez** [comment]
> ~~Still need to try without my "cache warming" script thingy.~~ I can confirm that this fixes it.

Turns out I had to _also_ disable my "do not update if nothing's changed" mu-plugin. I'll have to update it, take tags along when calculating the "hash."

I think what happens is the block editor really only saves the tags during an "update," i.e., after the post is first published a split second before, and that sending _only_ that first "Create" activity isn't enough to get tags to federate.

So, I won't reopen the issue, but think it's good to be aware nonetheless!

Another thing that _might_ solve it, is scheduling the activities a *bit* farther into the future. A number of seconds or so.

**Comment by janboddez** [comment]
Interestingly, I used to occasionally see the "Edited" flag when viewing posts on Mastodon, even when the post itself showed no difference whatsoever.

Howver, during this latest test, even though the tags became visible on Mastodon only after I'd clicked Update, the post in question does not at all show that is was updated.

Could be I saw that behavior when I was still relying on a cron job to process scheduled events, I can't say. Not sure if it could be related (at all).
